realtek: rtl930x: Add support for Plasma Cloud PSX8+PSX10 Switch

sharadanand opened the pull request:
The Plasma Cloud PSX8 Switch is a 8 port multi-GBit switch with 8x 10/100/1000/2500BaseT Ethernet ports. The PSX10 has additionally 2x SFP+ module slots.

Hardware:

- RTL9302C SoC
- Macronix MX25L25645G (32MB flash)
- Winbond W632GU6NB-12 (256MB DDR3 SDRAM - only 128 MB configured*)
- 2x RTL8224 4x 10m/100m/1/2.5 Gigabit PHY
- SFP+ 2x 10GBit slot (only on PSX10)
- IC+ IP8008 POE+ PSE controller

The switch is powered by 54 Volts 2.77A barrel connector. The internal TTL serial connector can be used to access the terminal. Pins from 1: TX RX (unused) GND.  Serial connection is via 115200 baud, 8N1.

A reset button is accessible through a hole in the front panel.

*) Only 128 MB of RAM are currently configured because there were infrequent random memory corruptions detected when using memtester with a 256 MB DT configuration. This could also be reproduced with RTLSDK.

Installation
------------

* The device can be flashed by using sysupgrade command. Either from the original vendor firmware or using an initramfs (see "Debug")
* Connect serial as per the layout above. Connection parameters: 115200 8N1
* The image (here `psx10`) must be copied using scp to /tmp of the device

      scp openwrt-realtek-rtl930x-plasmacloud_psx10-squashfs-sysupgrade.bin root@[IP address of the device]:/tmp/

* start sysupgrade without saving the original vendor configuration

      sysupgrade -n /tmp/openwrt-realtek-rtl930x-plasmacloud_psx10-squashfs-sysupgrade.bin

Installation via u-boot
-----------------------

If you have an TFTP server connected to the switch, it is possible to directly install the device using the factory image from u-boot

    # setup networking and IP of TFP server
    rtk network on
    setenv ipaddr 10.100.100.99
    setenv serverip 10.100.100.20
    
    # get factory image
    tftp 0x84000000 factory.bin
    
    # erase firmware partitions
    sf probe 0
    sf erase 0x100000 0x01f00000
    
    # write firmware to both partitions
    sf write ${fileaddr} 0x100000 ${filesize}
    sf write ${fileaddr} 0x1080000 ${filesize}
    
    # adjust the boot commands
    setenv bootargs "mtdparts=spi0.0:896k(u-boot),64k(u-boot-env),64k(u-boot-env2),15872k(inactive),15872k(firmware2)"
    setenv bootcmd "rtk init; bootm 0xb5080000"
    
    # restart
    reset

Debug
-----

* Connect serial as per the layout above. Connection parameters: 115200 8N1.
* A tftp server is required, tftpd-hpa works well.
* Power the device, at U-Boot start rapidly hit Esc key to stop autoboot
* Enable network:

      rtk network on

* Change ip address of device:

      setenv ipaddr 192.168.1.6

* Download initramfs from TFTP server:

      tftpboot 0x84000000 192.168.1.111:openwrt-realtek-rtl930x-plasmacloud_psx10-initramfs-kernel.bin

* Boot loaded file:

      bootm 0x84000000

0xharshal committed:
realtek: Fix nvmem support for ethernet device

The function of_get_mac_address is not taking care of evaluation the nvmem
address before trying to read out the mac-address properties. The driver
must check whether the return code is -EPROBE_DEFER and stop the probing
process in that case. If the nvmem-cell related driver code finished, the
probe can be redone ad the correct mac-address will appear for the device.

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19362
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: rtl930x: Enable parsing of u-boot nvmem layouts

To be able to read out the ethaddr from the u-boot environment for MAC
address configuration, it is required to also enable the NVMEM layout
parsing code for the U-Boot env layout.

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19362
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv reviewed (changes requested):
Regarding the commit message + the PR title+message: it is better than #19173 but is still not following the expected style.

Please have a look at commits like one in this list and update both commit and PR:

* 6af9476b8a3c75ce65f17cfa89da7d35360a4756
* 39b9b491bb585ba6c16dfc28e5dd0316c8fe7b0d
* ba674fea40260b71e4e479cb81d2ce96d5a85f73
* d9f30b64ad93d946452dcd69f6c54417f549654b

And again, similar to the other PR, you still didn't take the time to prepare the text for the PR:

> Thanks for your contribution to OpenWrt!
> 
> To help keep the codebase consistent and readable,
> and to help people review your contribution,
> we ask you to follow the rules you find in the wiki at this link
> https://openwrt.org/submitting-patches
> 
> Please remove this message before posting the pull request.

ecsv commented on target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot.sh:
```diff
@@ -0,0 +1,109 @@
+	[ -n "$UPGRADE_BACKUP" ] && restore_backup="${MTD_CONFIG_ARGS} -j ${UPGRADE_BACKUP}"
+
+	mtd -q erase $PART_NAME
+	tar -xf $tar_file ${board_dir}/kernel ${board_dir}/root -O | mtd $restore_backup write - $PART_NAME
```
The `mtd erase` right before the `mtd $restore_backup write - $PART_NAME` should not be needed. The `mtd write` itself will erase the relevant blocks. This will only slow down sysupgrade process.

sharadanand replied:
Done

ecsv commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,254 @@
+			label = "green:power";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_POWER;
+			gpios = <&gpio0 0 GPIO_ACTIVE_LOW>;
```
Can you please try to add

```
			linux,default-trigger = "heartbeat";
```

Instead of using `01_leds`

0xharshal commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,254 @@
+		led-boot = &led_power;
+		led-running = &led_power;
+		led-failsafe = &led_power;
+		led-upgrade = &led_power;
```
This should be read from the u-boot-env so this doesn't make sense to be part of device-tree.

@sharadanand I suggest using `sysinfo_fixup` in `/lib/preinit/` to set MAC correctly on boot

ecsv replied:
I would prefer when this could be tested first with the `nvmem-layout` in the devicetree.

ecsv replied:
According to https://github.com/openwrt/openwrt/pull/19173#discussion_r2200368674, there is a problem in the driver. This has to be tested - if it doesn't work correctly with nvmem, there is still the fallback to use `target/linux/realtek/base-files/etc/board.d/02_network` mac addresses.

ecsv commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,254 @@
+	};
+};
+
+&ethernet0 {
```
It looks to me like the mac address is not set in there. I should be done using:

```
&ethernet0 {
	nvmem-cells = <&macaddr_ubootenv_ethaddr>;
	nvmem-cell-names = "mac-address";
```


The `macaddr_ubootenv_ethaddr` should then be retrieved using a `nvmem-layout` sub-node under your `partition@e0000` node:

```
				nvmem-layout {
					compatible = "u-boot,env";

					macaddr_ubootenv_ethaddr: ethaddr {
						#nvmem-cell-cells = <1>;
					};
				};
```

ecsv replied:
According to https://github.com/openwrt/openwrt/pull/19173#discussion_r2200368674, there is a problem in the driver. This has to be tested - if it doesn't work correctly with nvmem, there is still the fallback to use `target/linux/realtek/base-files/etc/board.d/02_network` mac addresses.

ecsv commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,254 @@
+			color = <LED_COLOR_ID_RED>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
It is still in `01_leds`

ecsv commented on target/linux/realtek/base-files/etc/board.d/01_leds:
```diff
@@ -1,13 +0,0 @@
-
```
Why are you removing this file in this PR?

sharadanand replied:
That an accident will fix it :)

0xharshal committed:
realtek: Introduce Plasma Cloud sysupgrade helper

Plasma Cloud devices use a dual-firmware regions/slots boot mechanism. On
APs, the u-boot is "intelligent" and checks the NOR/NAND partitions (kernel
+ rootfs) for corruption via "datachk". If validation fails, the bootloader
automatically switches to the fallback partition.

On Realtek-based switches, this "datachk" helper is not available.
However, the bootloader still supports two firmware regions/slots.

When flashing a new image, the "inactive" partition is written instead of
overwriting the active one. If no "inactive" partition exists but
"firmware1" is present, the bootloader always treats "firmware1" as
fallback. Only after a successful flash is the `u-boot-env` updated to
select the newly written partition.

On reboot, the bootloader loads the kernel from the new partition and
passes `mtdparts` information as the kernel cmdline. The Plasma Cloud
switch device tree does not override this with a `bootargs` property, so
the active partition layout is honored from cmdline.

Since offsets, sizes, and names of partitions match between the device tree
and cmdline (except the inactive slot), properties and nodes such as
`nvmem-cells` or `compatible` remain fully usable.

This mechanism also allows switching back to the old firmware slot.  For
example, if `firmware1` is currently active (`/proc/mtd` shows it), it can
be switched to slot 2 using:

    . /lib/upgrade/upgrade_dualboot.sh
    set_boot_part 2
    reboot

Firmware upgrades use standard `sysupgrade` tarballs, chosen for
compatibility with vendor images. In theory, one can switch between vendor
and OpenWrt with:

    sysupgrade -n /tmp/*-squashfs-sysupgrade.bin

Note: configuration files must not be preserved, as they are not compatible
with vanilla OpenWrt.

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19362
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: rtl930x: Add support for Plasma Cloud PSX10 Switch

The Plasma Cloud PSX10 Switch is a 8 + 2 port multi-GBit switch with
8x 10/100/1000/2500BaseT Ethernet ports and 2x SFP+ module slot.

Hardware:

- RTL9302C SoC
- Macronix MX25L25645G (32MB flash)
- Winbond W632GU6NB-12 (256MB DDR3 SDRAM - only 128 MB configured*)
- 2x RTL8224 4x 10m/100m/1/2.5 Gigabit PHY
- SFP+ 2x 10GBit slot
- IC+ IP8008 POE+ PSE controller

The switch is powered by 54 Volts 2.77A barrel connector. The internal TTL
serial connector can be used to access the terminal. Pins from 1: TX RX
(unused) GND.  Serial connection is via 115200 baud, 8N1.

A reset button is accessible through a hole in the front panel.

*) Only 128 MB of RAM are currently configured because there were
infrequent random memory corruptions detected when using memtester with a
256 MB DT configuration. This could also be reproduced with RTLSDK.

Installation
------------

* The device can be flashed by using sysupgrade command. Either from the
  original vendor firmware or using an initramfs (see "Debug")
* Connect serial as per the layout above. Connection parameters: 115200 8N1
* The image must be copied using scp to /tmp of the device

      scp openwrt-realtek-rtl930x-plasmacloud_psx10-squashfs-sysupgrade.bin root@[IP address of the device]:/tmp/

* start sysupgrade without saving the original vendor configuration

      sysupgrade -n /tmp/openwrt-realtek-rtl930x-plasmacloud_psx10-squashfs-sysupgrade.bin

Installation via u-boot
-----------------------

If you have an TFTP server connected to the switch, it is possible to
directly install the device using the factory image from u-boot

    # setup networking and IP of TFP server
    rtk network on
    setenv ipaddr 10.100.100.99
    setenv serverip 10.100.100.20

    # get factory image
    tftp 0x84000000 factory.bin

    # erase firmware partitions
    sf probe 0
    sf erase 0x100000 0x01f00000

    # write firmware to both partitions
    sf write ${fileaddr} 0x100000 ${filesize}
    sf write ${fileaddr} 0x1080000 ${filesize}

    # adjust the boot commands
    setenv bootargs "mtdparts=spi0.0:896k(u-boot),64k(u-boot-env),64k(u-boot-env2),15872k(inactive),15872k(firmware2)"
    setenv bootcmd "rtk init; bootm 0xb5080000"

    # restart
    reset

Debug
-----

* Connect serial as per the layout above. Connection parameters: 115200 8N1.
* A tftp server is required, tftpd-hpa works well.
* Power the device, at U-Boot start rapidly hit Esc key to stop autoboot
* Enable network:

      rtk network on

* Change ip address of device:

      setenv ipaddr 192.168.1.6

* Download initramfs from TFTP server:

      tftpboot 0x84000000 192.168.1.111:openwrt-realtek-rtl930x-plasmacloud_psx10-initramfs-kernel.bin

* Boot loaded file:

      bootm 0x84000000

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Co-developed-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19362
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

0xharshal committed:
realtek: rtl930x: Add support for Plasma Cloud PSX8 Switch

The Plasma Cloud PSX8 Switch is a 8 port multi-GBit switch with
8x 10/100/1000/2500BaseT Ethernet ports.

Hardware:

- RTL9302C SoC
- Macronix MX25L25645G (32MB flash)
- Winbond W632GU6NB-12 (256MB DDR3 SDRAM - only 128 MB configured*)
- 2x RTL8224 4x 10m/100m/1/2.5 Gigabit PHY
- IC+ IP8008 POE+ PSE controller

The switch is powered by 54 Volts 2.77A barrel connector. The internal TTL
serial connector can be used to access the terminal. Pins from 1: TX RX
(unused) GND.  Serial connection is via 115200 baud, 8N1.

A reset button is accessible through a hole in the front panel.

*) Only 128 MB of RAM are currently configured because there were
infrequent random memory corruptions detected when using memtester with a
256 MB DT configuration. This could also be reproduced with RTLSDK.

Installation
------------

* The device can be flashed by using sysupgrade command. Either from the
  original vendor firmware or using an initramfs (see "Debug")
* Connect serial as per the layout above. Connection parameters: 115200 8N1
* The image must be copied using scp to /tmp of the device

      scp openwrt-realtek-rtl930x-plasmacloud_psx8-squashfs-sysupgrade.bin root@[IP address of the device]:/tmp/

* start sysupgrade without saving the original vendor configuration

      sysupgrade -n /tmp/openwrt-realtek-rtl930x-plasmacloud_psx8-squashfs-sysupgrade.bin

Installation via u-boot
-----------------------

If you have an TFTP server connected to the switch, it is possible to
directly install the device using the factory image from u-boot

    # setup networking and IP of TFP server
    rtk network on
    setenv ipaddr 10.100.100.99
    setenv serverip 10.100.100.20

    # get factory image
    tftp 0x84000000 factory.bin

    # erase firmware partitions
    sf probe 0
    sf erase 0x100000 0x01f00000

    # write firmware to both partitions
    sf write ${fileaddr} 0x100000 ${filesize}
    sf write ${fileaddr} 0x1080000 ${filesize}

    # adjust the boot commands
    setenv bootargs "mtdparts=spi0.0:896k(u-boot),64k(u-boot-env),64k(u-boot-env2),15872k(inactive),15872k(firmware2)"
    setenv bootcmd "rtk init; bootm 0xb5080000"

    # restart
    reset

Debug
-----

* Connect serial as per the layout above. Connection parameters: 115200 8N1.
* A tftp server is required, tftpd-hpa works well.
* Power the device, at U-Boot start rapidly hit Esc key to stop autoboot
* Enable network:

      rtk network on

* Change ip address of device:

      setenv ipaddr 192.168.1.6

* Download initramfs from TFTP server:

      tftpboot 0x84000000 192.168.1.111:openwrt-realtek-rtl930x-plasmacloud_psx8-initramfs-kernel.bin

* Boot loaded file:

      bootm 0x84000000

Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>
Co-developed-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Signed-off-by: Sharadanand Karanjkar <sk@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/19362
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv commented on target/linux/realtek/base-files/lib/upgrade/platform.sh:
```diff
@@ -35,5 +35,14 @@ platform_do_upgrade() {
+		PART_NAME="inactive"
+		platform_do_upgrade_dualboot_datachk "$1"
+		;;
+	*)
```
The default case must be always at the end - this part is correct. But you've duplicated the default case and forgot to remove the one before ` plasmacloud,psx8|\`

ecsv commented on target/linux/realtek/image/rtl930x.mk:
```diff
@@ -63,3 +63,34 @@ define Device/zyxel_xgs1250-12
+    uImage lzma
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma | pad-to $$(BLOCKSIZE)
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-kernel | append-rootfs | pad-rootfs
```
Nothing seems to use the factory image. So it doesn't need to be build.

ecsv commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,261 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		/* External RTL8224 PHY */
```
Support for this PHY seems to be missing. But there was a dubious patch in #19173

ecsv commented on target/linux/realtek/base-files/etc/board.d/02_network:
```diff
@@ -67,6 +67,16 @@ xikestor,sks8300-8x)
+	lan_mac="$label_mac"
+	# every port is assigned its own MAC
+	lan_mac_start=$(macaddr_add $lan_mac 1)
+	lan_mac_end=$(macaddr_add $lan_mac_start 28)
```
Neither PSX8 nor PSX10 have 28 ports

ecsv commented on target/linux/realtek/image/rtl930x.mk:
```diff
@@ -63,3 +63,34 @@ define Device/zyxel_xgs1250-12
+  SOC := rtl9302
+  UIMAGE_MAGIC := 0x93000000
+  DEVICE_VENDOR := Plasma Cloud
+  IMAGE_SIZE := 15872k
```
`check-size` missing:

```
IMAGE/factory.bin := append-kernel | append-rootfs | pad-rootfs | check-size
```

I am not sure if there was ever anything added for `sysupgrade-tar`. Because it is necessary to check here for both kernel + rootfs and not only the rootfs (when you add it after `pad-rootfs`) or the whole tar file (when you add it after `sysupgrade-tar`).

hailfinger commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,261 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		/* External RTL8224 PHY */
```
There also has been some progress on a spearate RTL8224 support effort: https://forum.openwrt.org/t/working-on-the-support-for-zyxel-xmg1915-10e/225898
Having RTL8224 support as a separate PR might help consolidate those efforts.

ecsv replied:
@hailfinger There is an upstream Linux RTL8224 driver since 9e42a2ea7f67 ("net: phy: realtek: add support for rtl8224 2.5Gbps PHY"). It is already part of Linux 6.11. It is built by OpenWrt and can be used directly by devices like the PSX8/PSX10.

Having a separate driver (and deactivating the upstream one), like it was done in https://github.com/openwrt/openwrt/compare/main...ddejean:openwrt:xmg1915-10e-wip2 (more specific https://github.com/openwrt/openwrt/commit/60d56911642a314ac65c2552f9ccaeb30c8a91b1 ), is not helping in the long run and might even break devices which depend on the upstream realtek phy driver support.

If there are problem with the upstream Linux drivers then they should be fixed in the Linux kernel code - preferably by also sending these changes directly to the Linux kernel maintainers.

---

Overall, this PR and the previous PR is misrepresenting work done by other people. And I am confused by @sharadanand's overall interactions. I will try to directly get in contact with him next week to get this figured out and hopefully get this and the promised other PRs in a state which can be reviewed by OpenWrt maintainers.

ecsv commented on target/linux/realtek/rtl930x/config-6.12:
```diff
@@ -123,7 +123,7 @@ CONFIG_MIPS=y
 CONFIG_MIPS_ASID_SHIFT=0
 # CONFIG_MIPS_CMDLINE_FROM_BOOTLOADER is not set
-CONFIG_MIPS_CMDLINE_FROM_DTB=y
+CONFIG_MIPS_CMDLINE_DTB_EXTEND=y
```
On ath79, this was done by simply removing the bootargs on devices which got this information from u-boot:

```
	chosen {
		/delete-property/ bootargs;
	};
```

have to check what/if there was a problem with realtek MIPS

Rondom replied:
How would that work? /delete-property/ removes a previously defined property at device-tree-compile-time.

ecsv replied:
> How would that work? /delete-property/ removes a previously defined property at device-tree-compile-time.

1. it works
2. @Rondom  `CONFIG_MIPS_CMDLINE_FROM_DTB` just just enables code which overwrites the cmdline with the content of `bootargs`. if it is not present, it will simply not overwrite anything. You can find the code which implements this behavior at https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/mips/kernel/setup.c?id=37816488247ddddbc3de113c78c83572274b1e2e#n588

ecsv commented on target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot.sh:
```diff
@@ -0,0 +1,79 @@
+	board=$(board_name)
+
+	case "$board" in
+		plasmacloud,psx8|\
+		plasmacloud,psx10)
+			if [ "$part_num" = "1" ]; then
+				echo "bootargs mtdparts=spi0.0:896k(u-boot),64k(u-boot-env),64k(u-boot-env2),15872k(firmware1),15872k(inactive)" > $setenv_script
+				echo "bootcmd rtk init; bootm 0xb4100000" >> $setenv_script
+			elif [ "$part_num" = "2" ]; then
+				echo "bootargs mtdparts=spi0.0:896k(u-boot),64k(u-boot-env),64k(u-boot-env2),15872k(inactive),15872k(firmware2)" > $setenv_script
+				echo "bootcmd rtk init; bootm 0xb5080000" >> $setenv_script
+			else
+				echo "Partition number $part_num is not supported for ${board}" 2>&1
+				return 1
+			fi
+			;;
+		*)
+			echo "${board} is not supported for dual boot" 1>&2
+			return 1
+			;;
```
Nitpicking: remove one indentation level

ecsv commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,265 @@
+				compatible = "openwrt,uimage", "denx,uimage";
+				openwrt,ih-magic = <0x93000000>;
+			};
+			partition@0x1080000 {
```
```suggestion
			partition@1080000 {
```

ecsv commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,271 @@
+			reg = <8>;
+			label = "lan5";
+			phy-mode = "usxgmii";
+			phy-handle = <&phy8>;
+			led-set = <0>;
+		};
+
+		port@9 {
+			reg = <9>;
+			label = "lan6";
+			phy-mode = "usxgmii";
+			phy-handle = <&phy9>;
+			led-set = <0>;
+		};
+
+		port@10 {
+			reg = <10>;
+			label = "lan7";
+			phy-mode = "usxgmii";
+			phy-handle = <&phy10>;
+			led-set = <0>;
+		};
+
+		port@11 {
+			reg = <11>;
+			label = "lan8";
+			phy-mode = "usxgmii";
+			phy-handle = <&phy11>;
+			led-set = <0>;
+		};
+
+		port@28 {
+			ethernet = <&ethernet0>;
+			reg = <28>;
+			phy-mode = "internal";
+			fixed-link {
+				speed = <1000>;
+				full-duplex;
+			};
+		};
```
When https://github.com/openwrt/openwrt/pull/19772 is merged earlier ("led-set" can be skipped because it is 0 by default):

```suggestion
		SWITCH_PORT(0, 1, usxgmii)
		SWITCH_PORT(1, 2, usxgmii)
		SWITCH_PORT(2, 3, usxgmii)
		SWITCH_PORT(3, 4, usxgmii)
		SWITCH_PORT(8, 5, usxgmii)
		SWITCH_PORT(9, 6, usxgmii)
		SWITCH_PORT(10, 7, usxgmii)
		SWITCH_PORT(11, 8, usxgmii)

		/* CPU-port */
		port@28 {
			ethernet = <&ethernet0>;
			reg = <28>;
			phy-mode = "internal";
			fixed-link {
				speed = <1000>;
				full-duplex;
			};
		};
```

plappermaul commented:
@simonwunderlich Sorry to have no time to look at this. Refactoring is keeping me busy. Regarding new devices I am somehow passionless. To get this merged I advise to focus only on the PSX8 and drop the other commits for (LED should not be crucial). Split the rest into two parts

- upgrade/dualboot scripts: explain what they are doing and why it is helpful. (makes the review easier for me)
- DTS + recipe: That already looks good from a short look.

Thanks.

plappermaul commented on target/linux/realtek/dts/rtl9302_plasmacloud_psx8.dtsi:
```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "rtl930x.dtsi"
+
```
Please rename this DTS include to rtl9302_plasmacloud.dtsi or rtl9302_plasmacloud_common.dtsi. While it is basically the whole PSX8 file we never use dedicated model names in dtsi files. Instead provide a short and clean rtl9302_plasmacloud_psx8.dts (without the I) derived from that (like the PSX10).

plappermaul reviewed (changes requested):
Thanks for providing these new devices. Maybe I did not make clear in what order I wanted the patches. Shame on me. I required only two commits.

- One for the upgrade script
- One for PSX8

Now we have only PSX10 (but with a lot of PSX8 fragments). As the patches are otherwise quite ok, make three out of it.

- One for the upgrade script
- One for PSX8 (including all common and PSX8 specific scripts and the specific dts amd the common dtsi)
- One for PSX10 (including PSX10 specific dts + scripts and repeating the installation procedure "I assume it is the same")

ecsv commented on target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot.sh:
```diff
@@ -0,0 +1,61 @@
+	fi
+
+	if [ -z "$inactive_mtd" ]; then
+		echo "Cannot find firmware1 or inactive partition to work with. Abort!"
```
"firmware1" and "inactive" are the actual labels of the partitions. How can be something more vague like "primary partition" be better?

plappermaul commented:
> @simonwunderlich Sorry to have no time to look at this. Refactoring is keeping me busy. Regarding new devices I am somehow passionless. To get this merged I advise to focus only on the PSX8 and drop the other commits for (LED should not be crucial). Split the rest into two parts
> 
> * upgrade/dualboot scripts: explain what they are doing and why it is helpful. (makes the review easier for me)
> * DTS + recipe: That already looks good from a short look.
> 
> Thanks.

That was what I requested. I only asked for PSX8 because I saw the close coupling of both switches. PSX8 being a subset of PSX10. Maybe I mumbled through a towel ... Now we have [realtek: rtl930x: Add support for Plasma Cloud PSX10 Switch](https://github.com/openwrt/openwrt/pull/19362/commits/7210f3e3278d4519d65dfa88b77243430ce5602d) which is 95% PSX8 and 5% PSX10. Why start with that? But who cares. No issues with that. Im fine with everything. If you want to start only with PSX10 then ok. 

Regarding the PSX8 dts. Yes. If this only consists of a single line then that is what we want here. E.g.

```
$ cat rtl8382_d-link_dgs-1210-28.dts
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT

#include "rtl838x.dtsi"
#include "rtl83xx_d-link_dgs-1210_common.dtsi"
#include "rtl83xx_d-link_dgs-1210_gpio.dtsi"
#include "rtl83xx_d-link_dgs-1210_gpio_sfp.dtsi"
#include "rtl8382_d-link_dgs-1210-28_common.dtsi"

/ {
        compatible = "d-link,dgs-1210-28", "realtek,rtl838x-soc";
        model = "D-Link DGS-1210-28";
};
```

ecsv commented:
> Regarding the PSX8 dts. Yes. If this only consists of a single line then that is what we want here. E.g.

This is clear to me. I was talking about the `realtek,forced-port-mask = <0x0c000f0f>;` and the sudden dependency to #19573

plappermaul reviewed (approved):
I assume that SWITCH_PORT(0, 1, usxgmii) is wrong and will need 10G_QSXGMII migration later if other patches are in place. For now I held back the USXGMII serdes patches to avoid tears. https://github.com/openwrt/openwrt/pull/19834#issuecomment-3219748376

Nevertheless I like.

hauke commented on package/boot/uboot-tools/uboot-envtools/files/realtek:
```diff
@@ -31,6 +31,15 @@ zyxel,gs1900-48)
 	;;
+plasmacloud,psx8|\
+plasmacloud,psx10)
+	idx="$(find_mtd_index u-boot-env)"
+	[ -n "$idx" ] && \
+		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x10000"
```
Use the new `ubootenv_add_mtd`:
```suggestion
	ubootenv_add_mtd "u-boot-env" "0x0" "0x10000"
```

hauke replied:
`ubootenv_add_uci_config` gets called with one more argument in most other places, does the u-bot configuration work?

ecsv replied:
Should most likely be dropped because it would match the default case:

```
*)
	ubootenv_add_mtd "u-boot-env" "0x0" "0x10000" "0x10000"
	ubootenv_add_sys_mtd "u-boot-env2" "0x0" "0x1000" "0x10000"
	;;
```

hauke commented on package/boot/uboot-tools/uboot-envtools/files/realtek:
```diff
@@ -31,6 +31,15 @@ zyxel,gs1900-48)
+	idx="$(find_mtd_index u-boot-env)"
+	[ -n "$idx" ] && \
+		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x10000"
+	idx2="$(find_mtd_index u-boot-env2)"
+	[ -n "$idx2" ] && \
+		ubootenv_add_uci_sys_config "/dev/mtd$idx2" "0x0" "0x1000"
```
Use the new `ubootenv_add_sys_mtd`:
```suggestion
	ubootenv_add_sys_mtd "u-boot-env2" "0x0" "0x10000"
```

ecsv replied:
Should most likely be dropped because it would match the default case:

```
*)
	ubootenv_add_mtd "u-boot-env" "0x0" "0x10000" "0x10000"
	ubootenv_add_sys_mtd "u-boot-env2" "0x0" "0x1000" "0x10000"
	;;
```

plappermaul commented:
@hauke thanks for the review.

ecsv commented on target/linux/realtek/dts/rtl9302_plasmacloud_common.dtsi:
```diff
@@ -0,0 +1,225 @@
+		phy0: ethernet-phy@0 {
+			reg = <0>;
+			compatible = "ethernet-phy-ieee802.3-c45";
+			rtl9300,smi-address = <0 0>;
```
This is the PHY, not the MAC.

You would need something like:

```
&lan1 {
	nvmem-cells = <&macaddr_ubootenv_ethaddr 1>;
	nvmem-cell-names = "mac-address";
};
```

ecsv commented on target/linux/realtek/base-files/etc/board.d/02_network:
```diff
@@ -47,6 +47,20 @@ hpe,1920-48g-poe)
 	lan_mac_end=$(macaddr_add $lan_mac $((mac_count2-mac_count1)))
 	;;
+plasmacloud,psx8)
+	label_mac=$(mtd_get_mac_ascii u-boot-env ethaddr)
```
Doesn't seem to work:

```
root@OpenWrt:~# fw_printenv ethaddr
ethaddr=54:9C:27:20:1F:50
root@OpenWrt:~# get_mac_label

```

ecsv replied:
Ok, `CONFIG_NVMEM_LAYOUT_U_BOOT_ENV=y` is not even set for this subtarget.

ecsv replied:
The DT only mac address configuration works with:

```diff
diff --git a/package/boot/uboot-tools/uboot-envtools/files/realtek b/package/boot/uboot-tools/uboot-envtools/files/realtek
index 2f0cd33f68..9b8bf90c27 100644
--- a/package/boot/uboot-tools/uboot-envtools/files/realtek
+++ b/package/boot/uboot-tools/uboot-envtools/files/realtek
@@ -31,15 +31,6 @@ zyxel,gs1900-48)
 	ubootenv_add_mtd "u-boot-env" "0x0" "0x400" "0x10000"
 	ubootenv_add_sys_mtd "u-boot-env2" "0x0" "0x1000" "0x10000"
 	;;
-plasmacloud,psx8|\
-plasmacloud,psx10)
-	idx="$(find_mtd_index u-boot-env)"
-	[ -n "$idx" ] && \
-		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x10000"
-	idx2="$(find_mtd_index u-boot-env2)"
-	[ -n "$idx2" ] && \
-		ubootenv_add_uci_sys_config "/dev/mtd$idx2" "0x0" "0x1000"
-	;;
 tplink,sg2008p-v1|\
 tplink,sg2210p-v3|\
 tplink,sg2452p-v4)
diff --git a/target/linux/realtek/base-files/etc/board.d/02_network b/target/linux/realtek/base-files/etc/board.d/02_network
index b641690957..9736f585a6 100644
--- a/target/linux/realtek/base-files/etc/board.d/02_network
+++ b/target/linux/realtek/base-files/etc/board.d/02_network
@@ -47,20 +47,8 @@ hpe,1920-48g-poe)
 	lan_mac_start=$(macaddr_add $lan_mac 2)
 	lan_mac_end=$(macaddr_add $lan_mac $((mac_count2-mac_count1)))
 	;;
-plasmacloud,psx8)
-	label_mac=$(mtd_get_mac_ascii u-boot-env ethaddr)
-	lan_mac="$label_mac"
-	# every port is assigned its own MAC
-	lan_mac_start=$(macaddr_add $lan_mac 1)
-	lan_mac_end=$(macaddr_add $lan_mac_start 8)
-	;;
-plasmacloud,psx10)
-	label_mac=$(mtd_get_mac_ascii u-boot-env ethaddr)
-	lan_mac="$label_mac"
-	# every port is assigned its own MAC
-	lan_mac_start=$(macaddr_add $lan_mac 1)
-	lan_mac_end=$(macaddr_add $lan_mac_start 10)
-	;;
+plasmacloud,psx8|\
+plasmacloud,psx10|\
 tplink,sg2008p-v1|\
 tplink,sg2210p-v3|\
 tplink,sg2452p-v4|\
diff --git a/target/linux/realtek/base-files/lib/upgrade/platform.sh b/target/linux/realtek/base-files/lib/upgrade/platform.sh
index 4a51769092..b4236cdf42 100644
--- a/target/linux/realtek/base-files/lib/upgrade/platform.sh
+++ b/target/linux/realtek/base-files/lib/upgrade/platform.sh
@@ -30,7 +30,7 @@ platform_do_upgrade() {
 	plasmacloud,psx8|\
 	plasmacloud,psx10)
 		PART_NAME="inactive"
-		platform_do_upgrade_dualboot_psx "$1"
+		platform_do_upgrade_dualboot_plasmacloud "$1"
 		;;
 	tplink,sg2008p-v1|\
 	tplink,sg2210p-v3)
diff --git a/target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot.sh b/target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot_plasmacloud.sh
similarity index 98%
rename from target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot.sh
rename to target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot_plasmacloud.sh
index 9640c10a24..4b390869e5 100644
--- a/target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot.sh
+++ b/target/linux/realtek/base-files/lib/upgrade/upgrade_dualboot_plasmacloud.sh
@@ -32,7 +32,7 @@ set_boot_part() {
 	}
 }
 
-platform_do_upgrade_dualboot_psx() {
+platform_do_upgrade_dualboot_plasmacloud() {
 	local primary_firmware_mtd
 	local restore_backup
 	local next_boot_part
diff --git a/target/linux/realtek/dts/rtl9302_plasmacloud_common.dtsi b/target/linux/realtek/dts/rtl9302_plasmacloud_common.dtsi
index c41ab1cee5..b20823f842 100644
--- a/target/linux/realtek/dts/rtl9302_plasmacloud_common.dtsi
+++ b/target/linux/realtek/dts/rtl9302_plasmacloud_common.dtsi
@@ -129,7 +129,7 @@
 };
 
 &ethernet0 {
-	nvmem-cells = <&macaddr_ubootenv_ethaddr>;
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 0>;
 	nvmem-cell-names = "mac-address";
 
 	mdio: mdio-bus {
@@ -223,3 +223,43 @@
 		};
 	};
 };
+
+&port0 {
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 1>;
+	nvmem-cell-names = "mac-address";
+};
+
+&port1 {
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 2>;
+	nvmem-cell-names = "mac-address";
+};
+
+&port2 {
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 3>;
+	nvmem-cell-names = "mac-address";
+};
+
+&port3 {
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 4>;
+	nvmem-cell-names = "mac-address";
+};
+
+&port8 {
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 5>;
+	nvmem-cell-names = "mac-address";
+};
+
+&port9 {
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 6>;
+	nvmem-cell-names = "mac-address";
+};
+
+&port10 {
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 7>;
+	nvmem-cell-names = "mac-address";
+};
+
+&port11 {
+	nvmem-cells = <&macaddr_ubootenv_ethaddr 8>;
+	nvmem-cell-names = "mac-address";
+};
diff --git a/target/linux/realtek/dts/rtl9302_plasmacloud_psx10.dts b/target/linux/realtek/dts/rtl9302_plasmacloud_psx10.dts
index d28b5827dd..b451bc24e3 100644
--- a/target/linux/realtek/dts/rtl9302_plasmacloud_psx10.dts
+++ b/target/linux/realtek/dts/rtl9302_plasmacloud_psx10.dts
@@ -50,6 +50,9 @@
 			sfp = <&sfp0>;
 			led-set = <0>;
 			managed = "in-band-status";
+
+			nvmem-cells = <&macaddr_ubootenv_ethaddr 9>;
+			nvmem-cell-names = "mac-address";
 		};
 
 		port@27 {
@@ -60,6 +63,9 @@
 			sfp = <&sfp1>;
 			led-set = <0>;
 			managed = "in-band-status";
+
+			nvmem-cells = <&macaddr_ubootenv_ethaddr 10>;
+			nvmem-cell-names = "mac-address";
 		};
 	};
 };
diff --git a/target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c b/target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c
index ddfdabacb6..635f77c2ec 100644
--- a/target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c
+++ b/target/linux/realtek/files-6.12/drivers/net/ethernet/rtl838x_eth.c
@@ -3470,7 +3470,10 @@ static int __init rtl838x_eth_probe(struct platform_device *pdev)
 	 * 1) from device tree data
 	 * 2) from internal registers set by bootloader
 	 */
-	of_get_mac_address(pdev->dev.of_node, mac_addr);
+	err = of_get_ethdev_address(pdev->dev.of_node, mac_addr);
+	if (err == -EPROBE_DEFER)
+		return err;
+
 	if (is_valid_ether_addr(mac_addr)) {
 		rtl838x_set_mac_hw(dev, mac_addr);
 	} else {
diff --git a/target/linux/realtek/rtl930x/config-6.12 b/target/linux/realtek/rtl930x/config-6.12
index 47779f151d..bbfd2fd7bd 100644
--- a/target/linux/realtek/rtl930x/config-6.12
+++ b/target/linux/realtek/rtl930x/config-6.12
@@ -167,6 +167,7 @@ CONFIG_NO_GENERIC_PCI_IOPORT_MAP=y
 CONFIG_NR_CPUS=2
 CONFIG_NVMEM=y
 CONFIG_NVMEM_LAYOUTS=y
+CONFIG_NVMEM_LAYOUT_U_BOOT_ENV=y
 CONFIG_OF=y
 CONFIG_OF_ADDRESS=y
 CONFIG_OF_EARLY_FLATTREE=y
```
